src/hdchain.cpp

```diff
@@ -0,0 +1,211 @@
+{
+    // mnemonic was not set, fail
+    if (vchMnemonic.empty())
+        return false;
+
+    ssMnemonicRet = SecureString(vchMnemonic.begin(), vchMnemonic.end());
+    ssMnemonicPassphraseRet = SecureString(vchMnemonicPassphrase.begin(), vchMnemonicPassphrase.end());
+
+    return true;
+}
+
+bool CHDChain::SetSeed(const SecureVector& vchSeedIn, bool fUpdateID)
+{
+    vchSeed = vchSeedIn;
+
+    if (fUpdateID) {
+        id = GetSeedHash();
+    }
+
+    return !IsNull();
+}
+
+SecureVector CHDChain::GetSeed() const
+{
+    return vchSeed;
+}
+
+uint256 CHDChain::GetSeedHash()
+{
+    return Hash(vchSeed.begin(), vchSeed.end());
+}
+
+void CHDChain::DeriveChildExtKey(uint32_t nAccountIndex, bool fInternal, uint32_t nChildIndex, CExtKey& extKeyRet)
+{
+    // Use BIP44 keypath scheme i.e. m / purpose' / coin_type' / account' / change / address_index
+    CExtKey masterKey;              //hd master key
+    CExtKey purposeKey;             //key at m/purpose'
+    CExtKey cointypeKey;            //key at m/purpose'/coin_type'
+    CExtKey accountKey;             //key at m/purpose'/coin_type'/account'
+    CExtKey changeKey;              //key at m/purpose'/coin_type'/account'/change
+    CExtKey childKey;               //key at m/purpose'/coin_type'/account'/change/address_index
+
+    masterKey.SetMaster(&vchSeed[0], vchSeed.size());
+
+    // Use hardened derivation for purpose, coin_type and account
+    // (keys >= 0x80000000 are hardened after bip32)
+
+    // derive m/purpose'
+    masterKey.Derive(purposeKey, 44 | 0x80000000);
+    // derive m/purpose'/coin_type'
+    purposeKey.Derive(cointypeKey, Params().ExtCoinType() | 0x80000000);
+    // derive m/purpose'/coin_type'/account'
+    cointypeKey.Derive(accountKey, nAccountIndex | 0x80000000);
+    // derive m/purpose'/coin_type'/account/change
+    accountKey.Derive(changeKey, fInternal ? 1 : 0);
+    // derive m/purpose'/coin_type'/account/change/address_index
+    changeKey.Derive(extKeyRet, nChildIndex);
+}
+
+void CHDChain::AddAccount()
+{
+    LOCK(cs_accounts);
+    mapAccounts.insert(std::pair<uint32_t, CHDAccount>(mapAccounts.size(), CHDAccount()));
+}
+
+bool CHDChain::GetAccount(uint32_t nAccountIndex, CHDAccount& hdAccountRet)
+{
+    LOCK(cs_accounts);
+    if (nAccountIndex > mapAccounts.size() - 1)
+        return false;
+    hdAccountRet = mapAccounts[nAccountIndex];
+    return true;
+}
+
+bool CHDChain::SetAccount(uint32_t nAccountIndex, const CHDAccount& hdAccount)
+{
+    LOCK(cs_accounts);
+    // can only replace existing accounts
+    if (nAccountIndex > mapAccounts.size() - 1)
+        return false;
+    mapAccounts[nAccountIndex] = hdAccount;
+    return true;
+}
+
+size_t CHDChain::CountAccounts()
+{
+    LOCK(cs_accounts);
+    return mapAccounts.size();
+}
+
+std::string CHDPubKey::GetKeyPath() const
+{
+    return strprintf("m/44'/%d'/%d'/%d/%d", Params().ExtCoinType(), nAccountIndex, nChangeIndex, extPubKey.nChild);
```

Comment:
nit: what is the meaning of 44 here? Better to avoid magic numbers.

Reply:
that is the purpose level, as defined by BIP44
https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki

Reply:
I think it would be nice to add a comment pointing there and/or to use a named const value.